### ISSUE TITLE
Improved support for projectile?

### ISSUE THREAD

**Issue by josteink** [OmniSharp/omnisharp-emacs#203]
Right now, I use projectile (with helm) to do most of my project management and navigation. I have keyboard bindings which works "everywhere" and is finger-memory at this point.

This works fine for small projects, using git as the project-container.

At work though, we have giant projects, using TFS as a source-control for multi-gigabyte branches and what not. In those cases using the file-system or TFS-root as a projectile index does not really make sense, nor is it efficient. Especially not when working with smaller sub-solutions within these giant branches.

omnisharp-emacs exposes the command `omnisharp-navigate-to-solution-file`, which let's you navigate to any file in the solution using `ido-mode`. This is better, but it still *differs* from my projectile+helm workflow, so it feels kind of awkward.

Would it be possible to "export" this index to projectile somehow, using omnisharp's solution-file knowledge as the base for projectile?

**Comment by josteink** [comment]
In projectile you have this function which I typically use: [projectile-find-file](https://github.com/bbatsov/projectile/blob/master/projectile.el#L1416). It  further digs into [projectile-current-project-files](https://github.com/bbatsov/projectile/blob/master/projectile.el#L1182).

From what I can tell this basically:

* checks if the index is already cached. If not, it builds it.
* then returns values from the cache.

A "dirty" fix would be simply populating the cache, but that can easily break for a million reasons (like a user-initiated cache-flush). It would definitely be nice if there was a better way. Quickly skimming the source doesn't show me any official way to extend it though.

@bbatsov: Do you have any feedback on this issue?

**Comment by sp3ctum** [comment]
For git projects, projectile allows using git as the source of included
files. I think it does something like `git ls-files` in the background.
Could you specify some `tf` command to do the same thing on projectile's
side?

2015-10-27 12:02 GMT+02:00 Jostein Kjønigsen <notifications@github.com>:

**Comment by josteink** [comment]
I think you read my bug-report a bit too quickly. :)

Typically TFS is often used for branches hosting 10 or 100s of solutions. Therefore using the TFS branch-root as projectile-root with a subsequent file-system is both inaccurate and suboptimal.

````bash
$ find . | wc -l
118968
````

Above is an example of one of these TFS branch-roots I work with. When looking up classes and source-files in a 100-class project I specifically do **not** want to use the TFS root as a projectile root. It's not even interesting and will probably be slower and more work than just doing a regular (non-projectile) `find-file`.

Using TFS as the method to resolve this also excludes all the solutions hosted outside source-control or outside known/compatible source-control systems.

The only reason git works so well is because projects using git often have source-control scoped correctly to that project only. This assumption can sadly not be easily transferred to other source-control systems.

`omnisharp` *knows* all the files I have in my solution, much better than a source-control/filesystem-based scanning/searching approach ever will. And we already have that file-list in elisp-accessible format in `omnisharp-emacs,` so I think a much better approach would be using that instead.

Which is why I adressed this issue in `omnisharp-emacs` and not projectile :)

**Comment by sp3ctum** [comment]
Ah yes, I see what you mean now. I too work with such TF branches but don't
use omnisharp-emacs in my work atm.

I think you have a good point.

2015-10-27 12:53 GMT+02:00 Jostein Kjønigsen <notifications@github.com>:

**Comment by nosami** [comment]
@jsteink - Would you be happy with a helm interface for `omnisharp-navigate-to-solution-file`?

I feel like `projectile` is skewing the conversation here

**Comment by josteink** [comment]
I (naively?) assumed plugging it into `projectile` might be the better option, because this *is* after all the sort of tasks which we have projectile for. Having it projectile-capable would also give us switching between projects using `projectile-switch-project`, possibly for free.

That would obviously be the "ideal" solution.

But if that turns out to be lots of work, and not so easy to do, because `projectile` lacks extension-points, I'm going to be pragmatic and say that a drop-in-replacement for `projectile-find-file` which I can re-bind in `csharp-mode` would work too.

Basically, whatever doesn't break my workflow. :)

**Comment by nosami** [comment]
I think that plugging it into projectile might not be the best solution for everyone. `omnisharp-navigate-to-solution-file` only returns C# files, whereas web developers work on many types of files.

This is why I think having 2 separate commands is possibly a better option.

**Comment by josteink** [comment]
Good observation.

I honestly think (and expect) that when I tell omnisharp go to any solution file, it will give me *any* solution-file, c# or not.

But that does indeed bring up an interesting dilemma...

I often work with XML-files, resource-files, mime-files, CSS-files... Lot's of things embedded in my solutions and projects not C# which still I would like to have solution-based navigating for.

But without some programming-language agnostic super-mode which extends projectile, that would mean very strange behaviour when jumping from a csharp-mode/omnisharp-mode buffer to for instance a CSS-file or XML-file.

For now a pragmatic approach is a helm-based C# file-navigator.

But the "ultimate" solution would IMO be having a couple of primitive functions bundled together in a projectile-mode extension:

````elisp
(defun omnisharp-is-file-part-of-solution-p (file)
  "Returns t if `FILE' is part of the currently open omnisharp solution."
  ;;
  )

(defun omnisharp-projectile-mode ()
  "Reconfigures projectile to use the omnisharp file index."
  ;;
  )

(add-hook 'fundamental-mode-hook
          #'(lambda ()
              (when (omnisharp-is-file-part-of-solution-p)
                (omnisharp-projectile-mode))))
````

I think if we don't want to end up piling hacks and workarounds on top of workarounds, finding a way to plug our index into projectile-mode is honestly the only viable answer.

**Comment by josteink** [comment]
I opened this issue hoping extending `projectile` would be a fairly trivial thing to do.

In retrospect, I see that it wasn't really created to allow this, and this makes implementing this functionality hard.

Because of that, I'm closing this one until I see a better possible solution.
